src/tad_dftd3/util.py

```diff
@@ -44,6 +44,55 @@ def real_triples(numbers: Tensor, diagonal: bool = False) -> Tensor:
     return mask
 
 
+# https://discuss.pytorch.org/t/efficient-distance-matrix-computation/9065/6
+def cdist_quadratic_expansion(
+    x: Tensor, y: Optional[Tensor] = None, p: int = 2
+) -> Tensor:
+    eps = torch.tensor(
+        torch.finfo(x.dtype).eps,
+        device=x.device,
+        dtype=x.dtype,
+    )
+
+    xnorm = torch.pow(x, p).sum(-1)
+    if y is None:
+        ynorm = xnorm
+    else:
+        ynorm = torch.pow(y, p).sum(-1)
+
+    n = xnorm.unsqueeze(-1) + ynorm.unsqueeze(-2)
+
+    # x @ y.mT
+    prod = torch.einsum("...ik, ...jk -> ...ij", x, y)
+
+    # sum of squared differences or L2-norm of differences
+    # important: remove negative values that give NaN in backward
+    return torch.sqrt(torch.clamp(n - 2.0 * prod, min=eps))
+
+
+def cdist_direct_expansion(x: Tensor, y: Optional[Tensor] = None, p: int = 2) -> Tensor:
+    eps = torch.tensor(
+        torch.finfo(x.dtype).eps,
+        device=x.device,
+        dtype=x.dtype,
+    )
+
+    if y is None:
+        y = x
+
+    # unsqueeze different dimension to create matrix
+    differences = x.unsqueeze(-2) - y.unsqueeze(-3)
+
+    distances = torch.sum(torch.pow(differences, p), -1)
+    return torch.sqrt(torch.clamp(distances, min=eps))
+
+
+def cdist(
+    x: Tensor, y: Optional[Tensor] = None, p: int = 2, **_: dict[str, Any]
```

Comment:
Why are additional arguments dropped here?

Reply:
I am currently slightly reworking this part, but I was hoping that this is just a temporary replacement, and we can use `torch.cdist` in the future again. So I did not want to replace all `torch.cdist` calls, which currently look like this:

```python
torch.cdist(positions, positions, p=2, compute_mode="use_mm_for_euclid_dist" )
```

But then, I do not think it is likely that PyTorch would include handling of zero division for the backward of `torch.cdist`. So I think we should switch to one of the manual implementations.

Reply:
I just realized that we could do something like:

```python
def cdist(x: Tensor, y: Optional[Tensor] = None, p: int = 2) -> Tensor:
  return torch.cdist(x, y, p=p, compute_mode="use_mm_for_euclid_dist")
```

if we wanted to switch back. So I will remove the unnecessary arguments.

Reply:
Maybe habe a look into `functools.partial` here, which might be clearer than creating thin wrapper functions

Reply:
I don't quite see, how this could make clearer. We will not use `torch.cdist` anymore, only `cdist`. And in the end, one should only call this wrapper.

Reply:
@awvwgk Do you still consider this necessary? Otherwise, I would like to merge and fix the release (the current one is broken because of the missing npy file).